cudax/include/cuda/experimental/__device/device_vector.cuh

```diff
@@ -0,0 +1,231 @@
+};
+
+//! @brief A random-access iterator over all available CUDA devices
+struct device_vector::iterator
+{
+  using value_type        = device;
+  using difference_type   = int;
+  using pointer           = const device*;
+  using reference         = device;
+  using iterator_category = ::cuda::std::random_access_iterator_tag;
+  using iterator_concept  = ::cuda::std::random_access_iterator_tag;
+
+  //! @brief Proxy object used as the return value of `operator->`
+  struct __proxy
+  {
+    _CCCL_NODISCARD const device* operator->() const noexcept
+    {
+      return &__dev_;
+    }
+
+    device __dev_;
+  };
+
+  _CCCL_NODISCARD constexpr device operator*() const noexcept
+  {
+    return device(__id_);
+  }
+
+  _CCCL_NODISCARD constexpr device operator[](_CUDA_VSTD::size_t __index) const noexcept
+  {
+    return device(__id_ + static_cast<int>(__index));
+  }
+
+  _CCCL_NODISCARD constexpr __proxy operator->() const noexcept
+  {
+    return __proxy{device(__id_)};
+  }
+
+  constexpr iterator& operator++() noexcept
+  {
+    ++__id_;
+    return *this;
+  }
+
+  constexpr iterator operator++(int) noexcept
+  {
+    auto tmp = *this;
+    ++__id_;
+    return tmp;
+  }
+
+  constexpr iterator& operator--() noexcept
+  {
+    --__id_;
+    return *this;
+  }
+
+  constexpr iterator operator--(int) noexcept
+  {
+    auto tmp = *this;
+    --__id_;
+    return tmp;
+  }
+
+  constexpr iterator& operator+=(int __offset) noexcept
+  {
+    __id_ += __offset;
+    return *this;
+  }
+
+  _CCCL_NODISCARD constexpr iterator operator+(int __offset) noexcept
+  {
+    return iterator(__id_ + __offset);
+  }
+
+  constexpr iterator& operator-=(int __offset) noexcept
+  {
+    __id_ -= __offset;
+    return *this;
+  }
+
+  _CCCL_NODISCARD constexpr iterator operator-(int __offset) noexcept
+  {
+    return iterator(__id_ - __offset);
+  }
+
+  _CCCL_NODISCARD_FRIEND constexpr iterator operator+(int __lhs, const iterator& __rhs) noexcept
+  {
+    return iterator(__lhs + __rhs.__id_);
+  }
+
+  _CCCL_NODISCARD_FRIEND constexpr int operator-(const iterator& __lhs, const iterator& __rhs) noexcept
+  {
+    return __lhs.__id_ - __rhs.__id_;
+  }
+
+  _CCCL_NODISCARD_FRIEND constexpr bool operator==(const iterator& __lhs, const iterator& __rhs) noexcept
+  {
+    return __lhs.__id_ == __rhs.__id_;
+  }
+
+  _CCCL_NODISCARD_FRIEND constexpr bool operator!=(const iterator& __lhs, const iterator& __rhs) noexcept
+  {
+    return __lhs.__id_ != __rhs.__id_;
+  }
+
+private:
+  friend struct device_vector;
+
+  constexpr explicit iterator(int i) noexcept
+      : __id_(i)
+  {}
+
+  int __id_ = 0;
+};
+
+_CCCL_NODISCARD inline constexpr device device_vector::operator[](_CUDA_VSTD::size_t __id_) const noexcept
+{
+  assert(__id_ < size());
+  return device(__id_);
+}
+
+_CCCL_NODISCARD inline _CUDA_VSTD::size_t device_vector::size() const
+{
+  int __count = 0;
+  _CCCL_TRY_CUDA_API(::cudaGetDeviceCount, "failed to get the count of CUDA devices", &__count);
```

Comment:
I will need a small amount of state per device to store the primary context. Once I have that change I will change this to just query the size of that state as a way to cache the device count

Reply:
for now i can cache the device count, and we can add per-device storage when we need that.